async.js

```diff
@@ -6,14 +7,58 @@
  */
 const isStar = true;
 
+/**
+ * Добавляет к Promise таймаут. Если таймаутзакончился, то вернется resolve(Error)
+ * @param {Number} ms - timeout
+ * @param {Promise} promise - promise
+ * @returns {Promise}
+ */
+function promiseTimeout(ms, promise) {
+    let timeout = new Promise((resolve, reject) => {
+        let id = setTimeout(() => {
+            clearTimeout(id);
+            reject(new Error('Promise timeout'));
+        }, ms);
+    });
+
+    return Promise.race([promise, timeout]);
+}
+
+
 /** Функция паралелльно запускает указанное число промисов
  * @param {Function<Promise>[]} jobs – функции, которые возвращают промисы
  * @param {Number} parallelNum - число одновременно исполняющихся промисов
  * @param {Number} timeout - таймаут работы промиса
  * @returns {Promise<Array>}
  */
 function runParallel(jobs, parallelNum, timeout = 1000) {
-    // асинхронная магия
+    if (jobs.length === 0) {
+        return Promise.resolve([]);
+    }
+
+    return new Promise(resolve => {
+        const results = new Array(jobs.length);
+        let finished = 0;
+        let count = 0;
```

Comment:
```suggestion
        let count = Math.min(parallelNum, jobs.length);
```
? 

А ниже как переменную цикла `i` использовать?